src/scanpy/preprocessing/_simple.py

```diff
@@ -732,13 +748,16 @@ def regress_out(
             )
             raise ValueError(msg)
         logg.debug("... regressing on per-gene means within categories")
-        regressors = np.zeros(X.shape, dtype="float32")
+
+        # set number of categories to the same dtype as the categories
+        cat_array = adata.obs[keys[0]].cat.codes.to_numpy()
+        number_categories = cat_array.dtype.type(len(adata.obs[keys[0]].cat.categories))
+
         X = _to_dense(X, order="F") if issparse(X) else X
-        # TODO figure out if we should use a numba kernel for this
-        for category in adata.obs[keys[0]].cat.categories:
-            mask = (category == adata.obs[keys[0]]).values
-            for ix, x in enumerate(X.T):
-                regressors[mask, ix] = x[mask].mean()
+        if np.issubdtype(X.dtype, np.integer):
+            target_dtype = np.float32 if X.dtype.itemsize == 4 else np.float64
```

Comment:
What's the point of `float32` here? Why not just always use `float64`? Wouldn't it be more accurate?

Reply:
Also, if there is a reason, just to be more clean:

```suggestion
            target_dtype = np.float32 if X.dtype.itemsize < 4 else np.float64
```

Reply:
Lastly, maybe I'm misremembering, but did we not discuss putting this in `to_dense` as well (so some sort of `target_dtype` created before the `X = _to_dense(X, order="F") if issparse(X) else X` line)? That way we don't have to make two copies in the sparse case.

Reply:
Should also apply to #3461